python/rrc_example_package/cube_env.py

```diff
@@ -0,0 +1,305 @@
+
+        if not self.action_space.contains(action):
+            raise ValueError(
+                "Given action is not contained in the action space."
+            )
+
+        num_steps = self.frameskip
+
+        # ensure episode length is not exceeded due to frameskip
+        step_count_after = self.step_count + num_steps
+        if step_count_after > move_cube.episode_length:
+            excess = step_count_after - move_cube.episode_length
+            num_steps = max(1, num_steps - excess)
+
+        reward = 0.0
+        for _ in range(num_steps):
+            self.step_count += 1
+            if self.step_count > move_cube.episode_length:
+                raise RuntimeError("Exceeded number of steps for one episode.")
+
+            # send action to robot
+            robot_action = self._gym_action_to_robot_action(action)
+            t = self.platform.append_desired_action(robot_action)
+
+            observation = self._create_observation(t)
+
+            reward += self.compute_reward(
+                observation["achieved_goal"],
+                observation["desired_goal"],
+                self.info,
+            )
+
+        self._previous_action = action
+
+        is_done = self.step_count == move_cube.episode_length
+
+        return observation, reward, is_done, self.info
+
+    def reset(self):
+        # reset is not really possible
+        if self.platform is not None:
+            raise RuntimeError(
+                "Once started, this environment cannot be reset."
+            )
+
+        self.platform = robot_fingers.TriFingerPlatformFrontend()
+        self.step_count = 0
+
+        # need to already do one step to get initial observation
+        # TODO disable frameskip here?
+        observation, _, _, _ = self.step(self._previous_action)
+
+        return observation
+
+    def seed(self, seed=None):
+        """Sets the seed for this env’s random number generator.
+
+        .. note::
+
+           Spaces need to be seeded separately.  E.g. if you want to sample
+           actions directly from the action space using
+           ``env.action_space.sample()`` you can set a seed there using
+           ``env.action_space.seed()``.
+
+        Returns:
+            List of seeds used by this environment.  This environment only uses
+            a single seed, so the list contains only one element.
+        """
+        self.np_random, seed = gym.utils.seeding.np_random(seed)
+        move_cube.random = self.np_random
+        return [seed]
+
+    def _create_observation(self, t):
+        robot_observation = self.platform.get_robot_observation(t)
+        camera_observation = self.platform.get_camera_observation(t)
+
+        observation = {
+            "observation": {
+                "position": robot_observation.position,
+                "velocity": robot_observation.velocity,
+                "torque": robot_observation.torque,
+            },
+            "previous_action": self._previous_action,
```

Comment:
i think this will be the wrong action. we want to return the latest action here, the one currently being executed. so i think actually we would want to call here self.platform.get_desired_action(t).

maybe you can double check yourself in the paper how we write to do it, to make sure we both understand it the same way.

Reply:
done